testkit/src/mill/testkit/BspServerUtil.scala

```diff
@@ -0,0 +1,184 @@
+          false,
+          if (exists) s"Error: value differs from fixture at $fixturePath"
+          else s"Error: no fixture found at $fixturePath"
+        )
+    else if (updateFixtures) {
+      // Fixture on disk might need to be updated anyway, if replaceAll changed
+      // and new strings should be replaced
+      val jsonStr = doReplaceAll(
+        gson.toJson(
+          value,
+          implicitly[ClassTag[T]].runtimeClass
+        )
+      )
+      val onDiskJsonStr = os.read(fixturePath)
+      if (jsonStr != onDiskJsonStr) {
+        System.err.println(s"Updating $fixturePath")
+        os.write.over(fixturePath, jsonStr)
+      }
+    }
+  }
+
+  val bspJsonrpcPool: ExecutorService = Executors.newCachedThreadPool(
+    new ThreadFactory {
+      val counter = new AtomicInteger
+      def newThread(runnable: Runnable): Thread = {
+        val t = new Thread(runnable, s"mill-bsp-integration-${counter.incrementAndGet()}")
+        t.setDaemon(true)
+        t
+      }
+    }
+  )
+
+  trait MillBuildServer extends b.BuildServer with b.JvmBuildServer
+      with b.JavaBuildServer with b.ScalaBuildServer
+
+  def withBspServer[T](
+      workspacePath: os.Path,
+      millTestSuiteEnv: Map[String, String],
+      client: b.BuildClient = dummyBuildClient
+  )(f: (MillBuildServer, b.InitializeBuildResult) => T): T = {
+
+    val jsonFile = workspacePath / Constants.bspDir / s"${Constants.serverName}.json"
+    assert(os.exists(jsonFile))
+    val contents = os.read(jsonFile)
+    assert(
+      !contents.contains("--debug"),
+      contents.contains(s""""bspVersion":"${bsp4jVersion}"""")
+    )
+
+    val contentsJson = ujson.read(contents)
+    val bspCommand = contentsJson("argv").arr.map(_.str)
+    val proc = os.proc(bspCommand).spawn(
+      cwd = workspacePath,
+      stderr = os.Inherit,
+      env = millTestSuiteEnv
+    )
+
+    try {
+      val launcher = new l.jsonrpc.Launcher.Builder[MillBuildServer]
+        .setExecutorService(bspJsonrpcPool)
+        .setInput(proc.stdout.wrapped)
+        .setOutput(proc.stdin.wrapped)
+        .setRemoteInterface(classOf[MillBuildServer])
+        .setLocalService(client)
+        .setExceptionHandler { t =>
+          System.err.println(s"Error during LSP processing: $t")
+          t.printStackTrace(System.err)
+          l.jsonrpc.RemoteEndpoint.DEFAULT_EXCEPTION_HANDLER.apply(t)
+        }
+        .create()
+
+      launcher.startListening()
+
+      val buildServer = launcher.getRemoteProxy()
+
+      val initRes = buildServer.buildInitialize(
+        new b.InitializeBuildParams(
+          "Mill Integration",
+          BuildInfo.millVersion,
+          b.Bsp4j.PROTOCOL_VERSION,
+          workspacePath.toNIO.toUri.toASCIIString,
+          new b.BuildClientCapabilities(List("java", "scala").asJava)
+        )
+      ).get()
+
+      f(buildServer, initRes)
+    } finally {
+      proc.stdin.close()
+      proc.stdout.close()
+
+      proc.join(2000L)
```

Comment:
2s seems a bit of an odd timeout; it's short enough that it could conceivably cause flakiness in test under CPU/memory/GC pressure, but long enough it won't during development when tests are typically run alone. If we just want to make sure it doesn't hang forever, maybe a longer timeout like 10s or 30s? That should be long enough it will only trigger if something is really-truly stuck.